qchem/pennylane_qchem/qchem/obs.py

```diff
@@ -0,0 +1,293 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This module contains functions to construct many-body observables whose expectation
+values can be used to simulate molecular properties.
+"""
+# pylint: disable=too-many-arguments, too-few-public-methods
+import os
+import numpy as np
+from openfermion.hamiltonians import MolecularData
+from openfermion.ops import FermionOperator
+from openfermion.transforms import bravyi_kitaev, jordan_wigner
+
+from . import structure
+
+
+def s2_me_table(sz, n_spin_orbs):
+    r"""Generates the table of the matrix elements
+    :math:`\langle \alpha, beta \vert \hat{s}_1 \cdot \hat{s}_2 \vert \gamma, \delta \rangle`
+    of the two-particle spin operator :math:`\hat{s}_1 \cdot \hat{s}_2`.
+
+    The matrix elements are evaluated using the expression,
+
+    .. math:
+
+        \langle \alpha, \beta \vert \hat{\bf{s}}_1 \cdot \hat {\bf{s}}_2
+        \vert \gamma, \delta \rangle = && \delta_{\alpha,\delta} \delta_{\beta,\gamma} \times \\
+        && \left( \frac{1}{2} \delta_{m_\alpha, m_\delta+1} \delta_{m_\beta, m_\gamma-1}
+        + \frac{1}{2} \delta_{m_\alpha, m_\delta-1} \delta_{m_\beta, m_\gamma+1}
+        + m_\alpha m_\beta \delta_{m_\alpha, m_\delta} \delta_{m_\beta, m_\gamma} \right),
+
+    where :math:`\alpha` and :math:`m_\alpha` refer to the quantum numbers of the spatial
+    :math:`\varphi_\alpha({\bf r})` and spin :math:`\chi_{m_\alpha}(s_z)` wave functions,
+    respectively, of the single-particle state :math:`\vert \alpha \rangle`.
+
+    **Example**
+
+    >>> sz = np.array([0.5, -0.5])
+    >>> print(s2_me_table(sz, 2))
+    [[ 0.    0.    0.    0.    0.25]
+     [ 0.    1.    1.    0.   -0.25]
+     [ 1.    0.    0.    1.   -0.25]
+     [ 1.    1.    1.    1.    0.25]
+     [ 0.    1.    0.    1.    0.5 ]
+     [ 1.    0.    1.    0.    0.5 ]]
+
+    Args:
+        sz (array[float]): spin-projection quantum number of the spin-orbitals
+        n_spin_orbs (int): number of spin orbitals
+
+    Returns:
+        array: NumPy array with the table of matrix elements
+    """
+
+    if sz.size != n_spin_orbs:
+        raise ValueError(
+            "Size of 'sz' must be equal to 'n_spin_orbs'; got {}".format(sz.size)
+        )
+
+    n = np.arange(n_spin_orbs)
+
+    alpha = n.reshape(-1, 1, 1, 1)
+    beta = n.reshape(1, -1, 1, 1)
+    gamma = n.reshape(1, 1, -1, 1)
+    delta = n.reshape(1, 1, 1, -1)
+
+    # we only care about indices satisfying the following boolean mask
+    mask = np.logical_and(alpha // 2 == delta // 2, beta // 2 == gamma // 2)
+
+    # diagonal elements
+    diag_mask = np.logical_and(sz[alpha] == sz[delta], sz[beta] == sz[gamma])
+    diag_indices = np.argwhere(np.logical_and(mask, diag_mask))
+    diag_values = (sz[alpha] * sz[beta]).flatten()
+
+    diag = np.vstack([diag_indices.T, diag_values]).T
```

Comment:
I have to admit that I always find very hard to read what all these reshaping does. I understand is faster but would not be able to tell you what happened in the last 15 lines.

Reply:
Only for @josh146 :-)

Reply:
No I agree, this function is very hard to read. I wonder how to improve it.